src/vsphere_cpi/lib/cloud/vsphere/resources/cluster.rb

```diff
@@ -45,12 +46,72 @@ def free_memory
         # Have to use separate mechanisms for fetching utilization depending on
         # whether we're using resource pools or raw clusters.
         if @config.resource_pool.nil?
-          @synced_free_memory = fetch_cluster_utilization
+          #$ Get cluster utilization only if host group is nil
+          @synced_free_memory = if host_group.nil?
+            fetch_cluster_utilization
+          else
+            fetch_host_group_utilization
+          end
         else
           @synced_free_memory = fetch_resource_pool_utilization
         end
       end
 
+      # Returns a vm host affinity rule name if specified in cluster config. Otherwise,
+      # it returns a default name made from host group and vm group.
+      #
+      # @return [String] Name of the host VM affinity rule name.
+      def vm_host_affinity_rule_name
+        # This first return might never get invoked as we always check host group
+        # to be present before calling this function.
+        # Just in case somebody uses this function in future without checking
+        # host group, this function will be safe.
+        return nil if host_group.nil?
+        return vm_group + '_' + host_group if config.vm_host_affinity_rule_name.nil?
+        config.vm_host_affinity_rule_name
+      end
+
+      # Returns a host group name if specified in cluster config. Otherwise,
+      # it returns a nil.
+      #
+      # @return [String] Name of the host Group specified under a cluster config.
+      def host_group
+        config.host_group
+      end
+
+      # Returns a vmgroup name if specified in cluster config. Otherwise,
+      # it returns a default name created by adding suffix to cluster name
+      #
+      # @return [String] Name of the VM Group specified under a cluster config.
+      def vm_group
+        return name + CLUSTER_VM_GROUP_SUFFIX if config.vm_group.nil?
+        config.vm_group
+      end
+
+      def host_group_mob
+        return nil if host_group.nil?
+        result_host_group = mob.configuration_ex.group.find do |group|
+          group.name == host_group && group.is_a?(VimSdk::Vim::Cluster::HostGroup)
+        end
+        raise "Failed to find the HostGroup:#{host_group} in Cluster:#{name}" if result_host_group.nil?
+        result_host_group
+      end
+
+      # Caches and returns the list of hosts under a cluster resource. It might
+      # be different from total hosts in a cluster if a host group is specified.
+      #
+      # In case a host group is present only the hosts under the host group are
+      # considered.
+      #
+      # If the host group does not exist/wrong it returns all the hosts
+      #
+      # @return [List] List of hosts in the cluster resource.
+      def host
```

Comment:
should be names hosts since it is returning a list of hosts and not a single host

Reply:
I agree. I will refactor this in another PR.  This was just kept host because of extensive usage and I did not want to change so many unrelated files.